File: python_implementation/dataframe.py
'''
Clase para manipular un data.frame personalizado con métodos para añadir y acceder columnas,
validar si está vacío, y exportar o importar datos en formatos CSV y TXT.
'''

class DataFrame:
    def __init__(self, datos: dict[str, list]):
        '''
        Constructor para inicializar el data.frame personalizado.
        Parameters:
            datos: (dict[str, list]) Diccionario que representa el data.frame, donde las llaves son nombres de columnas y los valores son listas con datos de cada columna.
        '''
        self.datos = datos
    
    def agregar_columna(self, nombre_columna: str, valores: list) -> None:
        '''
        Método para añadir una columna al data.frame.
        Parameters:
            nombre_columna: (str) Nombre de la nueva columna.
            valores: (list) Lista de valores para la columna.
        Return:
            None
        Exceptions:
            ValueError si la columna ya existe.
        '''
        if nombre_columna in self.datos:
            raise ValueError(f"La columna '{nombre_columna}' ya existe.")
        
        self.datos[nombre_columna] = valores
    
    def obtener_columna(self, nombre_columna: str) -> list:
        '''
        Método para obtener los valores de una columna específica.
        Parameters:
            nombre_columna: (str) Nombre de la columna a obtener.
        Return:
            list: Lista de valores de la columna solicitada.
        Exceptions:
            KeyError si la columna no existe.
        '''
        if nombre_columna not in self.datos:
            raise KeyError(f"La columna '{nombre_columna}' no existe.")
        
        return self.datos[nombre_columna]

    def esta_vacio(self) -> bool:
        '''
        Método para verificar si el data.frame está vacío.
        Return:
            bool: True si no hay datos o todas las columnas están vacías, False en caso contrario.
        '''
        return not bool(self.datos) or all(len(valores) == 0 for valores in self.datos.values())

    def __repr__(self) -> str:
        '''
        Método para representar el data.frame como una cadena de texto.
        Return:
            str: Representación visual del data.frame.
        '''
        filas = len(next(iter(self.datos.values()), []))
        return "\n".join(
            " | ".join(f"{self.datos[col][fila]}" for col in self.datos)
            for fila in range(filas)
        )
    
    def escribir_a_csv(self, ruta_archivo: str) -> None:
        '''
        Método para escribir el data.frame en un archivo CSV.
        Parameters:
            ruta_archivo: (str) Ruta donde se guardará el archivo CSV.
        Return:
            None
        '''
        with open(ruta_archivo, 'w') as archivo:
            archivo.write(','.join(self.datos.keys()) + '\n')
            filas = len(next(iter(self.datos.values()), []))
            for fila in range(filas):
                archivo.write(','.join(str(self.datos[col][fila]) for col in self.datos) + '\n')

    @classmethod
    def leer_desde_csv(cls, ruta_archivo: str) -> "DataFrame":
        '''
        Método de clase para leer un archivo CSV y crear un data.frame personalizado.
        Parameters:
            ruta_archivo: (str) Ruta del archivo CSV a leer.
        Return:
            DataFramePersonalizado: Nueva instancia de la clase con datos cargados desde el CSV.
        '''
        with open(ruta_archivo, 'r') as archivo:
            lineas = archivo.readlines()
            encabezados = lineas[0].strip().split(',')
            datos = {encabezado: [] for encabezado in encabezados}
            for linea in lineas[1:]:
                valores = linea.strip().split(',')
                for encabezado, valor in zip(encabezados, valores):
                    datos[encabezado].append(cls._convertir_tipo(valor))
            return cls(datos)

    def escribir_a_txt(self, ruta_archivo: str) -> None:
        '''
        Método para escribir el data.frame en un archivo TXT con valores separados por tabulaciones.
        Parameters:
            ruta_archivo: (str) Ruta donde se guardará el archivo TXT.
        Return:
            None
        '''
        with open(ruta_archivo, 'w') as archivo:
            archivo.write('\t'.join(self.datos.keys()) + '\n')
            filas = len(next(iter(self.datos.values()), []))
            for fila in range(filas):
                archivo.write('\t'.join(str(self.datos[col][fila]) for col in self.datos) + '\n')

    @classmethod
    def leer_desde_txt(cls, ruta_archivo: str) -> "DataFrame":
        '''
        Método de clase para leer un archivo TXT con valores separados por tabulaciones y crear un data.frame personalizado.
        Parameters:
            ruta_archivo: (str) Ruta del archivo TXT a leer.
        Return:
            DataFramePersonalizado: Nueva instancia de la clase con datos cargados desde el TXT.
        '''
        with open(ruta_archivo, 'r') as archivo:
            lineas = archivo.readlines()
            encabezados = lineas[0].strip().split('\t')
            datos = {encabezado: [] for encabezado in encabezados}
            for linea in lineas[1:]:
                valores = linea.strip().split('\t')
                for encabezado, valor in zip(encabezados, valores):
                    datos[encabezado].append(cls._convertir_tipo(valor))
            return cls(datos)

    @staticmethod
    def _convertir_tipo(valor: str) -> int | float | str:
        '''
        Método estático para convertir un valor en el tipo de dato adecuado (int, float o str).
        Parameters:
            valor: (str) Valor a convertir.
        Return:
            int | float | str: Valor convertido al tipo de dato adecuado.
        '''
        try:
            return int(valor)
        except ValueError:
            try:
                return float(valor)
            except ValueError:
                return valor

File: R_implementation/dataframe.R
#' Clase para manipular un data.frame personalizado con métodos para añadir y acceder columnas,
#' validar si está vacío, y exportar o importar datos en formatos CSV y TXT.
#'
#' @docType class
#' @name DataFrame
#' @export
DataFrame <- function(datos) {
  structure(list(datos = datos), class = "DataFrame")
}

#' Método para añadir una columna al data.frame.
#'
#' @param object Instancia del data.frame personalizado.
#' @param nombre_columna Nombre de la nueva columna.
#' @param valores Lista de valores para la columna.
#' @return El data.frame actualizado.
#' @examples
#' df <- DataFrame(list(col1 = c(1, 2, 3)))
#' df <- agregar_columna(df, "nueva_columna", c(4, 5, 6))
agregar_columna <- function(object, nombre_columna, valores) {
  if (nombre_columna %in% names(object$datos)) {
    stop(sprintf("La columna '%s' ya existe.", nombre_columna))
  }
  object$datos[[nombre_columna]] <- valores
  return(object)
}

#' Método para obtener los valores de una columna específica.
#'
#' @param object Instancia del data.frame personalizado.
#' @param nombre_columna Nombre de la columna a obtener.
#' @return Lista de valores de la columna solicitada.
#' @examples
#' obtener_columna(df, "col1")
obtener_columna <- function(object, nombre_columna) {
  if (!(nombre_columna %in% names(object$datos))) {
    stop(sprintf("La columna '%s' no existe.", nombre_columna))
  }
  return(object$datos[[nombre_columna]])
}

#' Método para verificar si el data.frame está vacío.
#'
#' @param object Instancia del data.frame personalizado.
#' @return TRUE si no hay datos o todas las columnas están vacías, FALSE en caso contrario.
#' @examples
#' esta_vacio(df)
esta_vacio <- function(object) {
  return(length(object$datos) == 0 || all(sapply(object$datos, function(valores) length(valores) == 0)))
}

#' Método para escribir el data.frame en un archivo CSV.
#'
#' @param object Instancia del data.frame personalizado.
#' @param ruta_archivo Ruta donde se guardará el archivo CSV.
#' @return None
#' @examples
#' escribir_a_csv(df, "datos.csv")
escribir_a_csv <- function(object, ruta_archivo) {
  datos_df <- as.data.frame(object$datos)
  write.csv(datos_df, ruta_archivo, row.names = FALSE)
}

#' Método para escribir el data.frame en un archivo TXT con valores separados por tabulaciones.
#'
#' @param object Instancia del data.frame personalizado.
#' @param ruta_archivo Ruta donde se guardará el archivo TXT.
#' @return None
#' @examples
#' escribir_a_txt(df, "datos.txt")
escribir_a_txt <- function(object, ruta_archivo) {
  datos_df <- as.data.frame(object$datos)
  write.table(datos_df, ruta_archivo, sep = "\t", row.names = FALSE, quote = FALSE)
}

#' Método para representar el data.frame como una cadena de texto.
#'
#' @param object Instancia del data.frame personalizado.
#' @return Cadena de texto que representa el data.frame.
#' @examples
#' print(df)
print.DataFrame <- function(object) {
  datos_df <- as.data.frame(object$datos)
  print(datos_df)
}

#' Método de clase para leer un archivo CSV y crear un data.frame personalizado.
#'
#' @param ruta_archivo Ruta del archivo CSV a leer.
#' @return Nueva instancia de la clase con datos cargados desde el CSV.
#' @examples
#' df <- leer_desde_csv("datos.csv")
leer_desde_csv <- function(ruta_archivo) {
  lineas <- readLines(ruta_archivo)
  encabezados <- strsplit(lineas[1], ",")[[1]]
  datos <- setNames(vector("list", length(encabezados)), encabezados)
  for (linea in lineas[-1]) {
    valores <- strsplit(linea, ",")[[1]]
    for (i in seq_along(encabezados)) {
      datos[[encabezados[i]]] <- c(datos[[encabezados[i]]], .convertir_tipo(valores[i]))
    }
  }
  DataFrame(datos)
}


#' Método de clase para leer un archivo TXT con valores separados por tabulaciones y crear un data.frame personalizado.
#'
#' @param ruta_archivo Ruta del archivo TXT a leer.
#' @return Nueva instancia de la clase con datos cargados desde el TXT.
#' @examples
#' df <- leer_desde_txt("datos.txt")
leer_desde_txt <- function(ruta_archivo) {
  lineas <- readLines(ruta_archivo)
  encabezados <- strsplit(lineas[1], "\t")[[1]]
  datos <- setNames(vector("list", length(encabezados)), encabezados)
  for (linea in lineas[-1]) {
    valores <- strsplit(linea, "\t")[[1]]
    for (i in seq_along(encabezados)) {
      datos[[encabezados[i]]] <- c(datos[[encabezados[i]]], .convertir_tipo(valores[i]))
    }
  }
  DataFrame(datos)
}

#' Método estático para convertir un valor en el tipo de dato adecuado (int, float o str).
#'
#' @param valor Valor a convertir.
#' @return Valor convertido al tipo de dato adecuado.
.convertir_tipo <- function(valor) {
  num_val <- suppressWarnings(as.numeric(valor))
  if (!is.na(num_val)) {
    if (num_val == as.integer(num_val)) {
      return(as.integer(num_val))
    } else {
      return(num_val)
    }
  } else {
    return(valor)
  }
}